controllers/vdoconfig_controller.go

```diff
@@ -668,86 +663,99 @@ func (r *VDOConfigReconciler) updateVdoConfigWithNodeStatus(ctx context.Context,
+					continue nodeloop
+				}
+				r.Logger.Info("adding node label", "name", VDO_NODE_LABEL_KEY, "node", node.Name)
 
-func (r *VDOConfigReconciler) reconcileNodeLabel(ctx vdocontext.VDOContext, req ctrl.Request, clientset *kubernetes.Clientset) error {
-	nodes, err := clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
-	if err != nil {
-		return errors.Wrapf(err, "Unable to fetch list of nodes")
-	}
+				node.Labels[VDO_NODE_LABEL_KEY] = req.Name
+				_, err := clientset.CoreV1().Nodes().Update(ctx, &node, metav1.UpdateOptions{})
 
-	for _, node := range nodes.Items {
-		if _, ok := node.Labels[VDO_NODE_LABEL_KEY]; !ok {
-			r.Logger.Info("adding node label", "name", VDO_NODE_LABEL_KEY)
-			node.Labels[VDO_NODE_LABEL_KEY] = req.Name
-			_, err = clientset.CoreV1().Nodes().Update(ctx, &node, metav1.UpdateOptions{})
-			if err != nil {
-				return errors.Wrapf(err, "Unable to update label on node")
+				if err != nil {
+					return errors.Wrapf(err, "Unable to update label on node")
+				}
 			}
 		}
 	}
 
 	return nil
 }
 
-func (r *VDOConfigReconciler) reconcileNodeProviderID(ctx vdocontext.VDOContext, config *vdov1alpha1.VDOConfig, clientset *kubernetes.Clientset) (*vdov1alpha1.VDOConfig, error) {
+func (r *VDOConfigReconciler) reconcileNodeProviderID(ctx vdocontext.VDOContext, config *vdov1alpha1.VDOConfig, clientset kubernetes.Interface, vsphereCloudConfigs *[]vdov1alpha1.VsphereCloudConfig) (*vdov1alpha1.VDOConfig, map[string]bool, error) {
 	nodeStatus := make(map[string]vdov1alpha1.NodeStatus)
-	cpiStatus := vdov1alpha1.Configured
 
+	cpiStatus := vdov1alpha1.Configured
 	nodes, err := clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
 	if err != nil {
-		return config, errors.Wrapf(err, "Unable to fetch list of nodes")
+		return config, NodeAvailabilityMap, errors.Wrapf(err, "Error reconciling the providerID for CPI, unable to fetch list of nodes")
 	}
 
+nodeLoop:
 	for _, node := range nodes.Items {
 		if len(node.Spec.ProviderID) > 0 {
 			nodeStatus[node.Name] = vdov1alpha1.NodeStatusReady
-		} else {
-			nodeStatus[node.Name] = vdov1alpha1.NodeStatusPending
-			cpiStatus = vdov1alpha1.Configuring
+			NodeAvailabilityMap[node.Name] = true
+			r.Logger.Info("Adding to Available nodes", "node", node.Name)
+			continue nodeLoop
+		}
+
+		for _, taint := range node.Spec.Taints {
+			if taint.Key == CLOUD_PROVIDER_INIT_TAINT_KEY {
+
+				if val, ok := NodeAvailabilityMap[node.Name]; ok {
+					if val {
+						nodeStatus[node.Name] = vdov1alpha1.NodeStatusPending
+						cpiStatus = vdov1alpha1.Configuring
+						r.Logger.Info("Adding to Available nodes", "node", node.Name)
+						continue nodeLoop
+					}
+				}
+
+				NodeAvailabilityMap[node.Name], err = r.checkNodeExistence(ctx, vsphereCloudConfigs, node)
+				if err != nil {
+					return config, NodeAvailabilityMap, errors.Wrapf(err, "Error reconciling the providerID for CPI")
+				}
+
+				if val, ok := NodeAvailabilityMap[node.Name]; ok {
+					if val {
```

Comment:
why not check for ok instead of val

Reply:
val will let us know if the NodeAvailabilityMap[node.Name] is true and then only we can proceed with adding the node to the NodeStatus Map. 
NodeAvailabilityMap will also contain the nodes that are not present in vsphereCloudConfig resource with status as "false"